### DISCUSSION TITLE
Stop send G code commands after a few minitues

### DISCUSSION THREAD

**Discussion by vilianch123** [arkypita/LaserGRBL#1498] [Q&A]
Hello,I use larserGRBL V4.3.0 to control my 2-axes controller. The connection is 232 serial and baudrate is 115200bps.
I use image2gcode tool to generate a g code file and I want to sent it to my controller.
But always everytime,I found larserGRBL would stop send G code again,so my axes would just stop moving.The only things
larserGRBL do is send a '?' to get controller's current status.
What's the main problem of this question？buffer size or mode to send characters?
Thank you in advance!

**Comment by arkypita** [comment]
Can you put pictures of controller, and screenshot of lasergrbl when hang?
PS. next time open thread under "discussion" instead of "issues" unless you're sure it's a lasergrbl bug.

**Reply by vilianch123** [reply]
Hello, all these attachments are what I have now. The txt file is captured from Bus Hound,**in line 16569**, laserGRBL send the last g code and stop send again. The zip file is my G code file from image2gcode tool.
**Is the wrong response from controller cause the laserGRBL stop send G code ?**

![lasergrbl stop here](https://user-images.githubusercontent.com/86226681/122768109-dcd8e100-d2d5-11eb-99af-292058b17742.PNG)
[lasergrbl_not_send_g_codes_1.txt](https://github.com/arkypita/LaserGRBL/files/6686843/lasergrbl_not_send_g_codes_1.txt)
[circle2.nc.zip](https://github.com/arkypita/LaserGRBL/files/6686847/circle2.nc.zip)

**Comment by arkypita** [comment]
It is evident, from the files you attached to me, that something is disturbing the communication between the PC and the engraver and vice versa. This causes the messages to arrive "broken" to LaserGRBL, which is no longer able to recognize and handle them correctly.

Among these there are some "ok" messages that the controller sends when it has received / interpreted a gcode line and which allow LaserGRBL to understand the line as "sent" and therefore to free the buffer counter of the corresponding number of characters.

![image](https://user-images.githubusercontent.com/8782035/122771924-2c53e880-d2a7-11eb-8eee-fcbd26da1cbd.png)

This can be understood from the pieces of message you see in the interface, and also from the communication log you have recorded.

Eg on line 16466 a "Run.WPos" message is received as "Ru..WPos" and the very next "ok" are blended inside the status report message that start with "<" and should stop at the next ">"

![image](https://user-images.githubusercontent.com/8782035/122772260-7a68ec00-d2a7-11eb-937e-2607a122a3cb.png)

This is a clear signal of electrical noise and you should read the laserGRBL faq because it has been fully explained as to be fixed hardware side.

**Reply by arkypita** [reply]
When noise prevent lasergrbl understand "ok" correctly, LaserGRBL believe that command are not received/handled by the machine, and so the buffer result full of chars. This prevent LaserGRBL sending new commands, so the process stuck,

LaserGRBL is also able to recover itself from this situation (that, I repeat, should be solved hardware side) but only if you have grbl setting $10=3

![image](https://user-images.githubusercontent.com/8782035/122774047-28c16100-d2a9-11eb-8fe1-3cd743d1c294.png)

**Reply by vilianch123** [reply]
Hello,arkypita,
this time,I have optimized my cnc controller,which could send response correctly.That means, when lasergrbl send
a ?,it would response completed "<xxx>",when a new g code is arrived,it would response with "ok".But still, lasergrbl would
stop send g code after a long time.The cnc file is the same as privious and I have a completed Bus Hund files captured as attachment. So what logic block sending g code instructions?Thank [you!
[Uploading test_record_stop.zip…]()

**Reply by arkypita** [reply]
> So what logic block sending g code instructions

The logic I already described to you is the only logic that can block LaserGRBL sending gcode.

The control board has an RX circular buffer with limited space of 127 character, LaserGRBL cannot send more than 127 character otherwise the previous command are overwritten/damaged.

LaserGRBL count the number of character sent, and keep count of the free space in RX buffer, and stop sends when the command that he need to send is bigger than the available (presumed) RX buffer space.

Each time an "ok" or "error" (followed by new line CR-LF) is received from LaserGRBL the oldest pending command is considered as managed and its length is decremented from the point of view of available RX buffer space.

This is called character-counting streaming protocol, and LaserGRBL implement it [as defined by grbl firmware](https://github.com/gnea/grbl/wiki/Grbl-v1.1-Interface#streaming-protocol-character-counting-recommended-with-reservation). 

There is no evidence that LaserGRBL does not do this job correctly, so the only reason it can cause the hang is that these ok messages get lost on the wire. By "lost on the wire" I mean that they are no longer recognizable in reception because they interlace and overlap or break up together with some other message.

In these cases LaserGRBL believes that certain commands are still in the buffer, and little by little the space that it believes to be available decreases. When it believes that there is no more space in the buffer LaserGRBL stops sending commands, you can understand that the situation is this because you find a series of about 20 commands with the clock icon (pending) and the indicator in the status bar of use of the buffer is full (you can read the number of used/available and free space on its tooltip).


PS. your attachment was not uploaded

**Reply by vilianch123** [reply]
[test_record_stop.zip](https://github.com/arkypita/LaserGRBL/files/6769066/test_record_stop.zip)
sorry,its here.

**Reply by arkypita** [reply]
A log file organized like this is very difficult to search into, but just at first look I can see some ok's not correctly sent/received:

Line 467025: ok (6f 6b) not followed by CR-LF (0d 0a)

![image](https://user-images.githubusercontent.com/8782035/124591021-17a55200-de5c-11eb-9033-86d2e8d1c2d9.png)

Can't say how many of that or similar events are in the file because it's difficult to search.

**Reply by arkypita** [reply]
I suggested you to set $10=3 to enable buffer size report in status report, and so enable LaserGRBL to auto-recover from this situation.

You can see if this is effective or not (so you have confirmation that this is the problem) by just looking into LaserGRBL log file and searching for "Handle Missing OK" message.

**Reply by arkypita** [reply]
what kind of controller do you have, and what is the firmware installed?

**Reply by vilianch123** [reply]
It's a controller,which ported grbl firmware into a stm32 platform and running in rtos!

**Reply by vilianch123** [reply]
Hello,arkypita.
Just another little question: If two or more "ok.."comes together and received meantime,just like this:"ok..ok..ok..".
Do lasergrbl think it as only one "ok.."responsed or?
I think there is condition that controller would send some "ok.."together and if program think it as only one "ok..",then
missing of "ok.."would happen.

**Reply by arkypita** [reply]
> Hello,arkypita.
> Just another little question: If two or more "ok.."comes together and received meantime,just like this:"ok..ok..ok..".
> Do lasergrbl think it as only one "ok.."responsed or?
> I think there is condition that controller would send some "ok.."together and if program think it as only one "ok..",then
> missing of "ok.."would happen.

LaserGRBL rx thread is working like this:

1. Read a line of data (read all characters from COM up to "newline") in a string
2. Process the string

![image](https://user-images.githubusercontent.com/8782035/125455044-266f8721-00cb-4f1b-adbf-1dcd9703fb5a.png)

Each "ok" is handled separately provided it is correctly terminated by a newline character, and it is not "melted" with some other message (i.e. inside a status message or having some weird chars in front of it). In other words, to be recognized correctly it must be preceded by a newline and followed by a newline.

With newline I mean \n or \r\n (\r is stripped out automatically)

**Reply by vilianch123** [reply]
So maybe the controller has buffered some "ok.."together with realtime status response and..
I mean I would optimized the controller and test it later.
Thank you!

**Reply by arkypita** [reply]
You can use internal logging feature of LaserGRBL to capture communication as seen by LaserGRBL
Menu ? activate extended log

**Comment by arkypita** [comment]
@vilianch123 please test latest 4.5.0 version. I add a workaround for intercept "broken ok" (not lost one!) and handle that as regular ok.

https://github.com/arkypita/LaserGRBL/releases/download/v4.5.0/install.exe

**Reply by arkypita** [reply]
Improvement in v4.6.0 currently pre-release, can be downloaded here for test: https://github.com/arkypita/LaserGRBL/releases/download/v4.6.0/install.exe
